crates/rover-client/build.rs

```diff
@@ -12,73 +13,101 @@ use uuid::Uuid;
-
-            let response = client
-                .head(schema_url)
-                .send()
-                .expect("Failed to get headers from Apollo's schema download url.");
-
-            let remote_etag = response.headers().get("etag").and_then(|v| v.to_str().ok());
-            eprintln!("remote etag: {}", remote_etag.unwrap_or("None"));
-
-            if let Some(remote_etag) = remote_etag {
-                if remote_etag == current_etag {
-                    eprintln!("etags match. Not updating schema.");
-                    return Ok(());
-                }
+            eprintln!(".schema/hash.id already exists");
+            let current_hash = String::from_utf8(read(hash_path)?).unwrap();
+            eprintln!("current hash: {}", current_hash);
+            let remote_hash = query_hash()?;
+
+            if remote_hash == current_hash {
+                eprintln!("hashes match. Not updating schema.");
+                return Ok(());
             }
         }
-        update_schema(&client, schema_url)
+        let (remote_hash, remote_schema) = query_schema_and_hash()?;
+        update_schema(&remote_hash, &remote_schema)
     } else {
         Ok(())
     }
 }
 
-fn update_schema(client: &Client, url: &str) -> std::io::Result<()> {
-    eprintln!("Updating schema.");
-    let response = client
-        .get(url)
-        .send()
-        .expect("Failed to get GraphQL schema from Apollo's schema download url.");
+fn query_hash() -> Result<String> {
+    let (hash, _) = query(false)?;
+    Ok(hash)
+}
 
-    let etag = response
-        .headers()
-        .get("etag")
-        .and_then(|v| v.to_str().ok())
-        .expect("Failed to get etag header from response.");
+fn query_schema_and_hash() -> Result<(String, String)> {
+    let (hash, schema) = query(true)?;
+    Ok((hash, schema.unwrap()))
+}
 
-    eprintln!("Saving {} to .schema/etag.id", etag);
-    write(".schema/etag.id", etag)?;
+fn update_schema(hash: &str, schema: &str) -> Result<()> {
+    eprintln!("Updating schema.");
 
-    let schema = response
-        .text()
-        .expect("Failed to get schema text from response.");
+    eprintln!("Saving {} to .schema/hash.id", hash);
+    write(".schema/hash.id", hash)?;
 
     eprintln!("Writing schema text to .schema/schema.graphql");
 
     write(".schema/schema.graphql", schema)?;
 
+    // old versions of Rover wrote to etag.id, since this is no longer needed,
+    // let's remove it from dev machines
+    let _ = fs::remove_file("./.schema/etag.id");
+
     Ok(())
 }
+
+const QUERY: &str = r#"query FetchSchema($fetchDocument: Boolean!) {
+  graph(id: "apollo-platform") {
+    variant(name: "current") {
+      latestPublication {
+        schema {
+          hash
+          document @include(if: $fetchDocument)
+        }
+      }
+    }
+  }
+}"#;
+
+fn query(fetch_document: bool) -> Result<(String, Option<String>)> {
+    let graphql_endpoint = option_env!("APOLLO_GRAPHQL_SCHEMA_URL")
```

Comment:
i'm not sure how useful this lever is - perhaps it would be better for us to support an environment variable that changes the _variant_ we're fetching the schema from? not sure! fwiw i'm not sure if anybody is even using this endpoint right now for their build setups - it's not super easy to prototype w/Rover quite yet because of the inflexibility of rust's graphql-client library

Reply:
Yea. I think for local dev stuff it would be nice if we could point to a different variant that has our testing or experimental features but it would require some work to make our testing variants public, so probably should scope out from this initial change?

This does set us up nicely to be able to fetch any schema from studio though so eventually devs could test things before they hit the public api.

Reply:
👍🏼 cool, i'll leave this as is then and as things progress we can revisit and make this a bit more configurable. thanks a bunch for doing the leg work to make this possible 😄